pkg/controller/kfservice/reconcilers/knative/configuration_reconciler.go

```diff
@@ -82,12 +84,31 @@ func (r *ConfigurationReconciler) reconcileDefault(kfsvc *v1alpha1.KFService) er
 }
 
 func (r *ConfigurationReconciler) reconcileCanary(kfsvc *v1alpha1.KFService) error {
+	canaryConfigurationName := constants.CanaryConfigurationName(kfsvc.Name)
 	if kfsvc.Spec.Canary == nil {
+		if kfsvc.Status.Canary.Name != "" {
```

Comment:
It's an antipattern to execute any behavior against the status of a k8s resource. Always read from the source of truth (i.e. etcd).

I would make another function called finalizeConfiguration that is executed if canary does not exist in the spec. Use client go to get it, and if it's already deleted, we're all good, otherwise, delete it.

You can think of finalize as the opposite of reconcile.

Reply:
This check is not actually needed though harmless. I just added to avoid one extra GET for every reconcile.  i will remove it

Reply:
Perhaps I havent' seen the updated code yet, but it would be great if this were all in a separate function for cleanliness.